File: src/main/java/io/dustin/finstage/financial/service/SymbolService.java
package io.dustin.finstage.financial.service;

import io.dustin.finstage.common.annotation.UseCase;
import io.dustin.finstage.common.response.PagedResponse;
import io.dustin.finstage.financial.domain.Symbol;
import io.dustin.finstage.financial.dto.ExternalSymbolListResponse;
import io.dustin.finstage.financial.dto.SymbolListRequest;
import io.dustin.finstage.financial.dto.SymbolResponse;
import io.dustin.finstage.financial.infra.client.SymbolClient;
import lombok.RequiredArgsConstructor;

import java.util.List;

@UseCase
@RequiredArgsConstructor
public class SymbolService {

    private final SymbolClient symbolClient;

    public PagedResponse<SymbolResponse> getSymbols(SymbolListRequest request) {
        int page = request.getPage();
        int size = request.getSize();

        ExternalSymbolListResponse external = symbolClient.fetchSymbols(page, size);

        List<SymbolResponse> items = external.getItems().stream()
                .map(item -> Symbol.of(item.getSymbol(), item.getName(), item.getCountry()))
                .map(SymbolResponse::from)
                .toList();

        return PagedResponse.<SymbolResponse>builder()
                .total(external.getTotal())
                .page(external.getPage())
                .size(external.getSize())
                .totalPages(external.getTotalPages())
                .hasNext(external.isHasNext())
                .hasPrev(external.isHasPrev())
                .items(items)
                .build();
    }
}

File: src/main/kotlin/io/dustin/finstage/financial/service/SymbolService.kt
package io.dustin.finstage.financial.service

import io.dustin.finstage.common.annotation.UseCase
import io.dustin.finstage.common.response.PagedResponse
import io.dustin.finstage.financial.domain.Symbol
import io.dustin.finstage.financial.dto.SymbolListRequest
import io.dustin.finstage.financial.dto.SymbolResponse
import io.dustin.finstage.financial.infra.client.SymbolClient

@UseCase
class SymbolService(
        private val symbolClient: SymbolClient
) {

    fun getSymbols(request: SymbolListRequest): PagedResponse<SymbolResponse> {
        val page = request.page
        val size = request.size

        val external = symbolClient.fetchSymbols(page, size)

        val items = external?.items?.map {
            val symbol = Symbol.of(it.symbol, it.name, it.country)
            SymbolResponse.from(symbol)
        } ?: emptyList()

        return PagedResponse(
                total = external?.total?.toLong() ?: 0,
                page = external?.page ?: page,
                size = external?.size ?: size,
                totalPages = external?.totalPages ?: 0,
                hasNext = external?.hasNext ?: false,
                hasPrev = external?.hasPrev ?: false,
                items = items
        )
    }
}
